### PR TITLE
New Traits

### PR DESCRIPTION
**Pull request by Scribbles0**
## About the PR

Three new traits are added to the game and alcohol is slightly more effective (raised from 2 to 4).
Paracusia trait (original trait by @areitpog): Auditory hallucinations. Plays imaginary sounds.
Lightweight Drunk trait (by me, lmk if you have a better name for this please): This trait doubles the "effectiveness" of alcohol, making it last longer. (i.e. the shader lasts twice as long)
Muted trait makes you unable to speak.

**Media**

- [X] I have added screenshots/videos to this PR showcasing its changes ingame, **or** this PR does not require an ingame showcase

Paracusia Demonstration (time between sounds decreased for demo):

https://user-images.githubusercontent.com/91828755/219294528-87e1133b-80b0-48ae-9c9f-fe2cdd5a53b1.mp4

Lightweight Drunk Demonstration: I don't have a video for this one as it'd be extremely long but I ran some tests and it works. If I still need some form of demonstration/explanation, let me know.

Muted Demonstration:

https://user-images.githubusercontent.com/91828755/218033099-f5a95d3e-81b2-4e83-b8f6-1dfc2c59efdb.mp4

**Changelog**

:cl:
- add: Three new traits have been added Lightweight Drunk, Paracusia, and Muted! Check their descriptions to see what they do.
- tweak: Alcohol is now slightly more effective.

### CODE REVIEW

File: Content.Server/Traits/Assorted/ParacusiaComponent.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
+namespace Content.Server.Traits.Assorted;
+
+[RegisterComponent]
+public sealed class ParacusiaComponent : Component
+{
```
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/ParacusiaComponent.cs:5]: This should be in shared and the client should handle it, server doesn't need to worry about hallucinations as it's wasted bandwidth and cycles.

File: Content.Server/Traits/Assorted/ParacusiaSystem.cs [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            if (paracusia.NextIncidentTime >= 0)
+                continue;
+
+            // Set the new time.
+            paracusia.NextIncidentTime +=
+                _random.NextFloat(paracusia.minTimeBetweenIncidents, paracusia.maxTimeBetweenIncidents);
+
+            List<string> sounds = paracusia.Sounds ?? new List<string> { };
+            if (paracusia.Sounds == null || paracusia.Sounds.Count == 0)
+                return;
```
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/ParacusiaSystem.cs:44]: You're returning inside of a loop but this also won't matter when moved to the client.

File: Content.Shared/Traits/Assorted/LightweightComponent.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+namespace Content.Shared.Traits.Assorted;
+
+/// <summary>
+/// Used for the lightweight trait.
+/// </summary>
+[RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Shared/Traits/Assorted/LightweightComponent.cs:6]: This comment should better explain that it's related to drunksystem and what it does, it also needs networking.

File: Content.Server/Traits/Assorted/ParacusiaComponent.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
+namespace Content.Server.Traits.Assorted;
+
+[RegisterComponent]
+public sealed class ParacusiaComponent : Component
```
Comment (metalgearsloth) [inline Content.Server/Traits/Assorted/ParacusiaComponent.cs:4]: This should explain it's for auditory hallucinations.

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

**Comment by Scribbles0** [comment]
i think that solves all of those?
if not lmk

**Comment by metalgearsloth** [comment]
Paracusia still needs moving to the client.

File: Resources/Locale/en-US/traits/traits.ftl [hunk 5]
```diff
@@ -10,3 +10,9 @@ trait-sneezing-name = Runny nose
 trait-sneezing-desc = You sneeze and cough uncontrollably
 
 permanent-blindness-trait-examined = [color=lightblue]{CAPITALIZE(POSS-ADJ($target))} eyes are glassy and unfocused. It doesn't seem like {SUBJECT($target)} can see you.[/color]
+
+trait-lightweight-name = Lightweight Drunk
+trait-lightweight-desc = Alcohol has a stronger effect on you\
+
+trait-paracusia-name = Paracusia
+trait-paracusia-desc = You hear sounds that aren't really there
```
Comment (AlexMorgan3817) [inline Resources/Locale/en-US/traits/traits.ftl:18]:
```suggestion
trait-paracusia-desc = You hear sounds that aren't really there

```

File: Resources/Prototypes/Traits/disabilities.yml [hunk 6]
```diff
@@ -22,3 +22,60 @@
...
+        - /Audio/Voice/Talk/lizard.ogg
+        - /Audio/Voice/Talk/pai.ogg
+        - /Audio/Voice/Talk/speak_1.ogg
+        - /Audio/Voice/Talk/speak_2_ask.ogg
+        - /Audio/Voice/Talk/speak_3_exclaim.ogg
+        - /Audio/Voice/Human/malescream_1.ogg
+        - /Audio/Voice/Human/malescream_6.ogg
+        - /Audio/Voice/Human/femalescream_2.ogg
+        - /Audio/Voice/Human/femalescream_4.ogg
+        - /Audio/Voice/Vox/shriek1.ogg
```
Comment (AlexMorgan3817) [inline Resources/Prototypes/Traits/disabilities.yml:81]:
```suggestion
        - /Audio/Voice/Vox/shriek1.ogg

```

File: Resources/Prototypes/Traits/inconveniences.yml [hunk 7]
```diff
@@ -12,3 +12,10 @@
         params:
           variation: 0.2
       timeBetweenIncidents: 0.3, 300
+
+- type: trait
+  id: LightweightDrunk
+  name: trait-lightweight-name
+  description: trait-lightweight-desc
+  components:
+    - type: Lightweight
```
Comment (AlexMorgan3817) [inline Resources/Prototypes/Traits/inconveniences.yml:21]:
```suggestion
    - type: Lightweight

```

**Comment by Scribbles0** [comment]
okay, i think that should actually fix everything
if i'm still doing something wrong send me a message

File: Content.Client/Traits/ParacusiaSystem.cs [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+            if (paracusia.NextIncidentTime >= 0)
+                continue;
+
+            // Set the new time.
+            paracusia.NextIncidentTime +=
+                _random.NextFloat(paracusia.minTimeBetweenIncidents, paracusia.maxTimeBetweenIncidents);
+
+            List<string> sounds = paracusia.Sounds ?? new List<string> { };
+            if (paracusia.Sounds == null || paracusia.Sounds.Count == 0)
+                return;
```
Comment (metalgearsloth) [inline Content.Client/Traits/ParacusiaSystem.cs:45]: This is still returning in a loop, though you should just check if the attached entity has paracusia anyway.

The accumulator also needs to be a timespan.

File: Resources/Prototypes/Traits/disabilities.yml [hunk 9]
```diff
@@ -22,3 +22,60 @@
...
+- type: trait
+  id: Paracusia
+  name: trait-paracusia-name
+  description: trait-paracusia-desc
+  components:
+    - type: Paracusia
+      minTimeBetweenIncidents: 0.1
+      maxTimeBetweenIncidents: 300
+      maxSoundDistance: 7
+      sounds:
```
Comment (metalgearsloth) [inline Resources/Prototypes/Traits/disabilities.yml:35]: This isn't a valid datafield on the component.

File: Content.Shared/Traits/Assorted/ParacusiaComponent.cs [hunk 10]
```diff
@@ -0,0 +1,30 @@
+namespace Content.Shared.Traits.Assorted;
+
+/// <summary>
+/// This component is used for paracusia, which causes auditory hallucinations.
+/// </summary>
+[RegisterComponent]
```
Comment (metalgearsloth) [inline Content.Shared/Traits/Assorted/ParacusiaComponent.cs:6]: This needs networking.

File: Content.Client/Traits/ParacusiaSystem.cs [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+            // Offset position where the sound is played
+            Vector2 randomOffset =
+            new Vector2
+            (
+                _random.NextFloat(-paracusia.MaxSoundDistance, paracusia.MaxSoundDistance),
+                _random.NextFloat(-paracusia.MaxSoundDistance, paracusia.MaxSoundDistance)
+            );
+
+            EntityCoordinates newCoords = Transform(paracusia.Owner).Coordinates
+                .Offset(randomOffset);
```
Comment (metalgearsloth) [inline Content.Client/Traits/ParacusiaSystem.cs:56]: Does this cause issues if they're not parented to the grid?

File: Content.Client/Traits/ParacusiaSystem.cs [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+
+            // funy camera shake
+            _camera.KickCamera(paracusia.Owner, randomOffset);
+
+            // Play the sound
+            int randomIndex = _random.Next(0, paracusia.Sounds.Count);
```
Comment (metalgearsloth) [inline Content.Client/Traits/ParacusiaSystem.cs:62]: Just use Pick<T>

**Review by metalgearsloth — CHANGES_REQUESTED** [review]

File: Content.Client/Traits/ParacusiaSystem.cs (hunk 11 above)
Comment (Scribbles0) [reply Content.Client/Traits/ParacusiaSystem.cs:54]: Doesn't cause issues for me, at least. Tested it by walking off a station, if that counts.

**Comment by Scribbles0** [comment]
alright, i THINK that's everything. if there's still stuff wrong, lmk once again

**Comment by Scribbles0** [comment]
actually hold on i found a bug. let me fix it first

**Comment by Scribbles0** [comment]
pr ready for re-review

**Comment by Scribbles0** [comment]
i have decided to let the other 3 trait ideas i wrote down to be in another pr
aside from that this pr is finally ready for re-review

**Comment by Scribbles0** [comment]
new SFX that have been added (if anyone wants to review):
station alert levels (blue and red)
station events (meteors, power outage, etc)
zombie SFX
minibomb countdown
other stuff that probably won't cause IC problems

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 13]
```diff
@@ -19,6 +20,9 @@ public abstract class SharedDrunkSystem : EntitySystem
         if (applySlur)
             _slurredSystem.DoSlur(uid, TimeSpan.FromSeconds(boozePower), status);
 
+        if (EntityManager.HasComponent<LightweightComponent>(uid))
+            boozePower = 8f;
```
Comment (moonheart08) [inline Content.Shared/Drunk/DrunkSystem.cs:24]: Should this not just be a multiplier.

**Review by moonheart08 — CHANGES_REQUESTED** [review]

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 14]
```diff
@@ -19,6 +20,9 @@ public abstract class SharedDrunkSystem : EntitySystem
         if (applySlur)
             _slurredSystem.DoSlur(uid, TimeSpan.FromSeconds(boozePower), status);
 
+        if (EntityManager.HasComponent<LightweightComponent>(uid))
```
Comment (DrSmugleaf) [inline Content.Shared/Drunk/DrunkSystem.cs:23]: You should use the proxy method here.

```suggestion
        if (HasComp<LightweightComponent>(uid))
```

File: Content.Shared/Traits/Assorted/LightweightComponent.cs [hunk 15]
```diff
@@ -0,0 +1,12 @@
+using Robust.Shared.GameStates;
+using System.ComponentModel.DataAnnotations;
+
+namespace Content.Shared.Traits.Assorted;
+
+/// <summary>
+/// Used for the lightweight trait. DrunkSystem will check for this component and modify the boozePower accordingly if it finds it.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
+public sealed class LightweightComponent : Component
```
Comment (DrSmugleaf) [inline Content.Shared/Traits/Assorted/LightweightComponent.cs:10]: This should be named LightweightDrunkComponent for clarity, same as the name for the trait

**Review by DrSmugleaf — DISMISSED** [review]

File: Content.Client/Traits/ParacusiaSystem.cs [hunk 16]
```diff
@@ -0,0 +1,73 @@
...
+        var localPlayer = _player.LocalPlayer?.ControlledEntity;
+
+        if (!TryComp<ParacusiaComponent>(localPlayer, out var paracusia))
+            return;
+
+        if (curTime < paracusia.NextIncidentTime)
+            return;
+
+        // Set the new time.
+        paracusia.NextIncidentTime = _timing.CurTime +
```
Comment (AJCM-git) [inline Content.Client/Traits/ParacusiaSystem.cs:53]: same

File: Content.Client/Traits/ParacusiaSystem.cs [hunk 17]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    public override void Update(float frameTime)
+    {
+        base.Update(frameTime);
+
+        if (!_timing.IsFirstTimePredicted)
+            return;
+
+        var curTime = _timing.CurTime;
```
Comment (AJCM-git) [inline Content.Client/Traits/ParacusiaSystem.cs:43]: move curtime var right before is used

File: Content.Shared/Traits/Assorted/ParacusiaComponent.cs [hunk 18]
```diff
@@ -0,0 +1,47 @@
...
+using System;
+using Robust.Shared.Serialization;
+
+namespace Content.Shared.Traits.Assorted;
+
+/// <summary>
+/// This component is used for paracusia, which causes auditory hallucinations.
+/// </summary>
+[RegisterComponent, NetworkedComponent]
+public sealed class ParacusiaComponent : Component
```
Comment (AJCM-git) [inline Content.Shared/Traits/Assorted/ParacusiaComponent.cs:12]: VV fields

File: Content.Shared/Traits/Assorted/ParacusiaComponent.cs [hunk 19]
```diff
@@ -0,0 +1,47 @@
...
+    public IPlayingAudioStream? Stream;
+}
+
+[Serializable, NetSerializable]
+public sealed class ParacusiaComponentState : ComponentState
+{
+    public float MaxTimeBetweenIncidents { get; init; }
+    public float MinTimeBetweenIncidents { get; init; }
+    public float MaxSoundDistance { get; init; }
+    public SoundSpecifier Sounds { get; init; } = default!;
```
Comment (AJCM-git) [inline Content.Shared/Traits/Assorted/ParacusiaComponent.cs:46]: make these fields, not properties and make a constructor

File: Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs [hunk 20]
```diff
@@ -0,0 +1,49 @@
...
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<ParacusiaComponent, ComponentGetState>(GetCompState);
+        SubscribeLocalEvent<ParacusiaComponent, ComponentHandleState>(HandleCompState);
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(ShutdownParacusia);
+    }
+    private void GetCompState(EntityUid uid, ParacusiaComponent component, ref ComponentGetState args)
+    {
+        args.State = new ParacusiaComponentState
```
Comment (AJCM-git) [inline Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs:21]: constructor

File: Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs [hunk 21]
```diff
@@ -0,0 +1,49 @@
...
+            MaxTimeBetweenIncidents = component.MaxTimeBetweenIncidents,
+            MinTimeBetweenIncidents = component.MinTimeBetweenIncidents,
+            MaxSoundDistance = component.MaxSoundDistance,
+            Sounds = component.Sounds,
+        };
+    }
+
+    private void HandleCompState(EntityUid uid, ParacusiaComponent component, ref ComponentHandleState args)
+    {
+        if (args.Current is not ParacusiaComponentState state) return;
```
Comment (AJCM-git) [inline Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs:32]: dont put the return in the same line as the condition

File: Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs [hunk 22]
```diff
@@ -0,0 +1,49 @@
...
+public abstract class SharedParacusiaSystem : EntitySystem
+{
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<ParacusiaComponent, ComponentGetState>(GetCompState);
+        SubscribeLocalEvent<ParacusiaComponent, ComponentHandleState>(HandleCompState);
+        SubscribeLocalEvent<RoundRestartCleanupEvent>(ShutdownParacusia);
+    }
+    private void GetCompState(EntityUid uid, ParacusiaComponent component, ref ComponentGetState args)
```
Comment (AJCM-git) [inline Content.Shared/Traits/Assorted/SharedParacusiaSystem.cs:19]: newline

File: Content.Shared/Drunk/DrunkSystem.cs [hunk 23]
```diff
@@ -19,6 +20,9 @@ public abstract class SharedDrunkSystem : EntitySystem
...
+            boozePower = boozePower * 2;
```
Comment (AJCM-git) [inline Content.Shared/Drunk/DrunkSystem.cs:24]: make the multiplier part of the component, also `*=` for a shorter assignment

**Review by AJCM-git — CHANGES_REQUESTED** [review]

File: Content.Client/Traits/ParacusiaSystem.cs (hunk 16 above)
Comment (Scribbles0) [reply Content.Client/Traits/ParacusiaSystem.cs:53]: tried to do the same thing for this, trait did not work afterwards

**Comment by Scribbles0** [comment]
alright that should be all reviews addressed, if anything is wrong let me know

**Review by AJCM-git — COMMENTED** [review]
bors ping

**Review by AJCM-git — APPROVED** [review]

**Pull request merged**